File: src/leetcode/java/two_pointers/PartitionList_86.java
package leetcode.java.two_pointers;

public class PartitionList_86 {
  public ListNode partition(ListNode head, int x) {
    if (head == null) return null;
    ListNode sHead = new ListNode(-1);
    ListNode bHead = new ListNode(-1);
    ListNode smaller = sHead;
    ListNode bigger = bHead;
    while (head != null) {
      if (head.val >= x) {
        bigger.next = head;
        bigger = bigger.next;
      } else {
        smaller.next = head;
        smaller = smaller.next;
      }
      head = head.next;
    }
  }

  class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
      val = x;
      next = null;
    }
  }
}

File: src/leetcode/python/two_pointers/PartitionList_86.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class PartitionList_86(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        if not head:
            return None
        shd = smaller = ListNode(-1)
        bhd = bigger = ListNode(-1)
        while head:
            if head.val >= x:
                bigger.next = head
                bigger = bigger.next
            else:
                smaller.next = head
                smaller = smaller.next
            head = head.next
        smaller.next = bhd.next
        bigger.next = None
        return shd.next
